rust_src/src/keyboard.rs

```diff
@@ -122,6 +123,36 @@ pub fn lucid_event_type_list_p(event: Option<LispCons>) -> bool {
     })
 }
 
+// _Noreturn will be added to prototype by make-docfile.
+pub fn quit_recursive_edit(val: LispObject) -> ! {
+    unsafe {
+        if command_loop_level > 0 || minibuf_level > 0 {
+            Fthrow(Qexit, val);
+        }
+
+        let msg = "No recursive edit is in progress";
+        xsignal!(
+            Quser_error,
+            ::remacs_sys::make_string(
+                msg.as_ptr() as *const ::libc::c_char,
+                msg.len() as ::libc::ptrdiff_t,
+            )
+        );
+    }
+}
+
```

Comment:
Both of these could be `-> !` as well. Does the `lisp_fn` macro support that currently?

Reply:
Yes, it does.

Reply:
@db48x I have tried it but it didn't work.

Reply:
What was the problem with it?

Reply:
Some error that indicates it isn't supported. Sorry I forgot to copy it. We also don't have an occurrance in our code base of `lisp_fn` with `-> !`.

Reply:
```

error[E0277]: the trait bound `lisp::LispObject: std::convert::From<!>` is not satisfied
   --> src/keyboard.rs:144:1
    |
144 | #[lisp_fn]
    | ^^^^^^^^^^ the trait `std::convert::From<!>` is not implemented for `lisp::LispObject`
    |
    = help: the following implementations were found:
              <lisp::LispObject as std::convert::From<()>>
              <lisp::LispObject as std::convert::From<bool>>
              <lisp::LispObject as std::convert::From<std::option::Option<T>>>
              <lisp::LispObject as std::convert::From<std::vec::Vec<lisp::LispObject>>>
            and 23 others
    = note: required by `std::convert::From::from`
```

Reply:
Heh, that's not an error message you see often. I did think it would work though.

Reply:
Well, this actually seems to work:

```
impl From<!> for LispObject {
    fn from(_v: !) -> Self {
        // I'm surprized that this works
        Qnil
    }
}
```